### PR TITLE
Address CVE-2020-1938

### PR DESCRIPTION
**Pull request by frankgh**
The AJP protocol is enabled by default in Tomcat versions < 7.0.100,
with the AJP connector listening in TCP port 8009 by default. This is
low impact on PXF because PXF should only be accessible from segments
collocated with the PXF Server. It is important to note that mitigation
is only required if an AJP port is accessible to untrusted users.

Tomcat 7.0.100 disables AJP protocol by default

### CODE REVIEW

File: server/pxf-service/src/main/java/org/greenplum/pxf/service/HttpRequestParser.java [hunk 1]
```diff
@@ -391,24 +394,51 @@ private int parsePositiveIntOrError(String s, String propName) {
...
+        RequestMap(MultivaluedMap<String, String> requestHeaders) throws UnsupportedEncodingException {
             super(String.CASE_INSENSITIVE_ORDER);
+
+            boolean decodeHeaderValue = false;
             for (Map.Entry<String, List<String>> entry : requestHeaders.entrySet()) {
-                List<String> values = entry.getValue();
-                if (values == null) continue;
+                if (StringUtils.equalsIgnoreCase(ENCODED_HEADER_VALUES_NAME, entry.getKey())) {
+                    String value = getValue(entry.getValue());
+                    decodeHeaderValue = StringUtils.equals("true", value);
```
Comment (oliverralbertini) [inline server/pxf-service/src/main/java/org/greenplum/pxf/service/HttpRequestParser.java]: Shouldn't we ignore case on the value as well?

File: server/pxf-service/src/main/java/org/greenplum/pxf/service/HttpRequestParser.java [hunk 2]
```diff
@@ -391,24 +394,51 @@ private int parsePositiveIntOrError(String s, String propName) {
         private static final String PROP_PREFIX = "X-GP-";
         private static final String USER_PROP_PREFIX = "X-GP-OPTIONS-";
         private static final String USER_PROP_PREFIX_LOWERCASE = "x-gp-options-";
+        private static final String ENCODED_HEADER_VALUES_NAME = PROP_PREFIX + "ENCODED-HEADER-VALUES";
 
-        RequestMap(MultivaluedMap<String, String> requestHeaders) {
+        RequestMap(MultivaluedMap<String, String> requestHeaders) throws UnsupportedEncodingException {
```
Comment (denalex) [inline server/pxf-service/src/main/java/org/greenplum/pxf/service/HttpRequestParser.java]: why throw this exception ? We can't recover from it, might as well throw RuntimeException and simplify method signatures

File: server/pxf-service/src/main/java/org/greenplum/pxf/service/HttpRequestParser.java [hunk 3]
```diff
@@ -391,24 +394,51 @@ private int parsePositiveIntOrError(String s, String propName) {
...
+            for (Map.Entry<String, List<String>> entry : requestHeaders.entrySet()) {
+                String value = getValue(entry.getValue());
                 if (value == null) continue;
 
-                // Converting to value UTF-8 encoding
                 String key = entry.getKey();
-                value = new String(value.getBytes(StandardCharsets.ISO_8859_1), StandardCharsets.UTF_8);
+                if (decodeHeaderValue && StringUtils.startsWithIgnoreCase(key, PROP_PREFIX)) {
+                    value = URLDecoder.decode(value, StandardCharsets.UTF_8.name());
+                }
```
Comment (denalex) [inline server/pxf-service/src/main/java/org/greenplum/pxf/service/HttpRequestParser.java]: should we have unit tests that test that we trigger the logic on the header value, case insensitive and that decoding is correct ?

Comment (frankgh) [reply server/pxf-service/src/main/java/org/greenplum/pxf/service/HttpRequestParser.java]: yes! definitely

File: server/pxf-service/src/main/java/org/greenplum/pxf/service/HttpRequestParser.java (hunk 2 above)
Comment (frankgh) [reply server/pxf-service/src/main/java/org/greenplum/pxf/service/HttpRequestParser.java]: yeah, good point

File: server/pxf-service/src/main/java/org/greenplum/pxf/service/HttpRequestParser.java (hunk 3 above)
Comment (frankgh) [reply server/pxf-service/src/main/java/org/greenplum/pxf/service/HttpRequestParser.java]: @denalex  I just added unit tests

File: server/pxf-service/src/main/java/org/greenplum/pxf/service/HttpRequestParser.java [hunk 4]
```diff
@@ -391,24 +393,55 @@ private int parsePositiveIntOrError(String s, String propName) {
...
+                if (StringUtils.equalsIgnoreCase(ENCODED_HEADER_VALUES_NAME, entry.getKey())) {
+                    String value = getValue(entry.getValue());
+                    decodeHeaderValue = StringUtils.equalsIgnoreCase("true", value);
+                    break;
+                }
+            }
 
-                String value = values.size() > 1 ? StringUtils.join(values, ",") : values.get(0);
+            for (Map.Entry<String, List<String>> entry : requestHeaders.entrySet()) {
+                String value = getValue(entry.getValue());
```
Comment (frankgh) [inline server/pxf-service/src/main/java/org/greenplum/pxf/service/HttpRequestParser.java]: @oliverralbertini the logic was moved to the `getValue` method

File: server/pxf-service/src/main/java/org/greenplum/pxf/service/HttpRequestParser.java (hunk 1 above)
Comment (frankgh) [reply server/pxf-service/src/main/java/org/greenplum/pxf/service/HttpRequestParser.java]: it is `equalsIgnoreCase` now

**Pull request merged**